pkg/kratos/service.go

```diff
@@ -522,10 +663,105 @@ func (s *Service) UpdateSettingsFlow(
 	return settingsFlow, nil, resp.Cookies(), nil
 }
 
+func peekRequestBody(r *http.Request) ([]byte, error) {
+	body := r.Body
+	if body == nil {
+		return nil, fmt.Errorf("request body is nil")
+	}
+
+	defer body.Close()
+
+	bodyBytes, err := io.ReadAll(body)
+	if err != nil {
+		return nil, err
+	}
+
+	r.Body = io.NopCloser(bytes.NewReader(bodyBytes))
+
+	return bodyBytes, nil
+}
+
+func kratosFlowMethod(r *http.Request) (string, error) {
+	bodyBytes, err := peekRequestBody(r)
+	if err != nil {
+		return "", err
+	}
+
+	var methodPayload methodOnly
+	if err = json.Unmarshal(bodyBytes, &methodPayload); err != nil {
+		return "", err
+	}
+
+	return methodPayload.Method, nil
+}
+
+func (s *Service) ParseRegistrationFlowMethodBody(r *http.Request) (*kClient.UpdateRegistrationFlowBody, error) {
+	method, err := kratosFlowMethod(r)
+	if err != nil {
+		return nil, err
+	}
+
+	var ret kClient.UpdateRegistrationFlowBody
+	switch method {
+	case "password":
+		var body kClient.UpdateRegistrationFlowWithPasswordMethod
+		if err := parseBody(r.Body, &body); err != nil {
+			return nil, err
+		}
+		ret = kClient.UpdateRegistrationFlowWithPasswordMethodAsUpdateRegistrationFlowBody(&body)
+
+	case "profile":
+		body, err := parseProfileBody(r.Body)
+		if err != nil {
+			return nil, err
+		}
+		ret = kClient.UpdateRegistrationFlowWithProfileMethodAsUpdateRegistrationFlowBody(body)
+
+	case "oidc":
+		var body kClient.UpdateRegistrationFlowWithOidcMethod
+		if err := parseBody(r.Body, &body); err != nil {
+			return nil, err
+		}
+		ret = kClient.UpdateRegistrationFlowWithOidcMethodAsUpdateRegistrationFlowBody(&body)
+	default:
+		return nil, fmt.Errorf("unknown method: %s", method)
+	}
+
+	return &ret, nil
+}
+
+func parseProfileBody(body io.ReadCloser) (*kClient.UpdateRegistrationFlowWithProfileMethod, error) {
```

Comment:
Some identity traits need to be treated as immutable. For example we shouldn't allow a user to change their email.

The changes for registration require that we enable the profile method in Kratos. We can limit the endpoint's functionality in the login UI, by having immutable traits. With the latest networking changes, we do not expose the Kratos endpoints, BUT people with access to the k8s network will still have access to that endpoint. I think that we should implement a kratos hook to prevent this on a Kratos level (obviously out of scope for this PR, but I am not sure if it should be part of this epic).